src/librustc_yk_sections/emit_tir.rs

```diff
@@ -508,3 +538,79 @@ impl<'tcx> ToPack<ykpack::UnOp> for (&ConvCx<'_, 'tcx, '_>, &UnOp) {
         }
     }
 }
+
+/// Constant -> Pack
+impl<'tcx> ToPack<ykpack::Constant> for (&ConvCx<'_, 'tcx, '_>, &Constant<'tcx>) {
+    fn to_pack(&mut self) -> ykpack::Constant {
+        let (ccx, cst) = self;
+        (*ccx, cst.ty, cst.literal).to_pack()
+    }
+}
+
+//fn bits_to_pack<T>(k
+
+/// Ty + LazyConst -> Pack
+impl<'tcx> ToPack<ykpack::Constant> for (&ConvCx<'_, 'tcx, '_>, Ty<'tcx>, &LazyConst<'tcx>) {
+    fn to_pack(&mut self) -> ykpack::Constant {
+        let (ccx, ty, lconst) = self;
+        if let LazyConst::Evaluated(cst) = lconst {
+            (*ccx, *ty, cst).to_pack()
+        } else {
+            ykpack::Constant::Unimplemented
+        }
+    }
+}
+
+/// Ty + Const -> Pack
+impl<'tcx> ToPack<ykpack::Constant> for (&ConvCx<'_, 'tcx, '_>, Ty<'tcx>, &Const<'tcx>) {
```

Comment:
Dumb question: why are we throwing around triples like this rather than having a named struct?

Reply:
OK, this is hard to explain.

In the beginning, we were passing around only pairs of `(ccx, X)`, where `X` is the thing we want to lower and `ccx` carries state, like the next free TIR variable number and the compiler's `tcx` (needed for various queries).

As we diverged from MIR further, the need arose to refer to things we saw earlier in the lowering chain. In these cases those pairs became triples.

For example, in the above, we need to know the type of the MIR `Const` we are converting so that we can do an appropriate lowering to one of our TIR constant enums. The type arose a few stages earlier in the chain, so we have to pass it down.

We could define structs, but I recommend we don't. To be consistent, we'd need one for every kind of thing we want to lower. We currently have 22 lowerings and more will come. I also think it will be harder to understand, so let's leave it be.

Reply:
I must admit, something here doesn't feel right to me. If you think about an RPython trace, it's very simple in terms of the variety of things that can be in it, and we're going to want something similar. I suspect we're passing too much of MIR's complexity through to the run-time. Warning: I might be very wrong.

Reply:
I don't think you are wrong, but I'm also not sure what that has to do with passing the triples down at compile time.

I think we can certainly simplify our TIR, but in stages. I was going to try and simplify assignments in the next PR.

Reply:
OK, fair enough. Let's make the next PR purely about simplification then?

Reply:
Sure.